File: maxSubArrayLen.java
import java.util.*;
class maxSubArrayLen {
    public static void main(String [] args) {
        int[] nums ={1,-1,5,-2,3}; 
        int k = 3;
        int prefixSum = 0;
        int longestSubarray = 0;
        HashMap<Integer, Integer> indices = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            prefixSum += nums[i];

            // Check if all of the numbers seen so far sum to k.
            if (prefixSum == k) {
                longestSubarray = i + 1;
            }

            // If any subarray seen so far sums to k, then
            // update the length of the longest_subarray. 
            if (indices.containsKey(prefixSum - k)) {
                longestSubarray = Math.max(longestSubarray, i - indices.get(prefixSum - k));
            }
            
            // Only add the current prefix_sum index pair to the 
            // map if the prefix_sum is not already in the map.
            if (!indices.containsKey(prefixSum)) {
                indices.put(prefixSum, i);
            }
        }
        
        System.out.println(longestSubarray);
    }
}

File: maxSubArrayLen.py
def maxSubArrayLen(nums, k): # TC O(n) // SC O(n)
    prefix_sum = longest_subarray = 0
    indices = {}

    for i, num in enumerate(nums):
        prefix_sum += num
        
        # Check if all of the numbers seen so far sum to k.
        if prefix_sum == k:
            longest_subarray = i + 1
            
        # If any subarray seen so far sums to k, then
        # update the length of the longest_subarray. 
        if prefix_sum - k in indices:
            longest_subarray = max(longest_subarray, i - indices[prefix_sum - k])
            
        # Only add the current prefix_sum index pair to the 
        # map if the prefix_sum is not already in the map.
        if prefix_sum not in indices:
            indices[prefix_sum] = i

    return longest_subarray

print(maxSubArrayLen([1,-1,5,-2,3],3)) 